### PR TITLE
Add support for PagerDuty API v2

### LINKED ISSUE #877: PagerDuty v1 API will be unsupported from February 2018
See:
https://v2.developer.pagerduty.com/v2/docs/api-v2-frequently-asked-questions

> API v1 will be supported until February 6, 2018

We should migrate to v2.

### PR DESCRIPTION
**Pull request by scarvel8**
This closes #877 and maintains backwards compatibility with current configurations

### CODE REVIEW

File: notify/impl.go [hunk 1]
```diff
@@ -438,18 +438,32 @@ const (
 )
 
 type pagerDutyMessage struct {
-	ServiceKey  string            `json:"service_key"`
-	IncidentKey string            `json:"incident_key"`
-	EventType   string            `json:"event_type"`
-	Description string            `json:"description"`
+	RoutingKey  string            `json:"routing_key,omitempty"`
+	ServiceKey  string            `json:"service_key,omitempty"`
+	DedupKey    string            `json:"dedup_key"`
```
Comment (brian-brazil) [inline notify/impl.go]: What is this and the incident key?

I suspect neither should be user configurable.

File: config/config_test.go [hunk 2]
```diff
@@ -276,16 +276,17 @@ func TestEmptyFieldsAndRegex(t *testing.T) {
...
-			VictorOpsAPIURL:  "https://alert.victorops.com/integrations/generic/20131114/alert/",
+			ResolveTimeout:    model.Duration(5 * time.Minute),
+			SMTPSmarthost:     "localhost:25",
+			SMTPFrom:          "alertmanager@example.org",
+			HipchatAuthToken:  "mysecret",
+			HipchatURL:        "https://hipchat.foobar.org/",
+			SlackAPIURL:       "mysecret",
+			SMTPRequireTLS:    true,
+			PagerdutyURL:      "https://events.pagerduty.com/v2/enqueue",
+			PagerdutySeverity: "error",
```
Comment (brian-brazil) [inline config/config_test.go]: We don't have similar settings as global config options for other notifiers.

File: notify/impl.go [hunk 3]
```diff
@@ -474,17 +488,41 @@ func (n *PagerDuty) Notify(ctx context.Context, as ...*types.Alert) (bool, error
...
+
+	if n.conf.ServiceKey != "" {
+		msg.ServiceKey = tmpl(string(n.conf.ServiceKey))
+		msg.EventType = eventType
+		msg.IncidentKey = hashKey(key)
+		msg.Description = tmpl(n.conf.Description)
+		msg.Details = details
+		n.conf.URL = "https://events.pagerduty.com/generic/2010-04-15/create_event.json"
+	} else {
+		msg.RoutingKey = tmpl(string(n.conf.RoutingKey))
```
Comment (josedonizetti) [inline notify/impl.go]: RoutingKey is an integration key provided by pagerduty, should not be processed like a template.

File: notify/impl.go [hunk 4]
```diff
@@ -474,17 +488,41 @@ func (n *PagerDuty) Notify(ctx context.Context, as ...*types.Alert) (bool, error
...
+		msg.EventAction = eventType
+		msg.DedupKey = hashKey(key)
+		if eventType == pagerDutyEventTrigger {
+			msgpayload := &pagerDutyPayload{
+				Summary:       tmpl(n.conf.Description),
+				Source:        n.conf.Client,
+				Severity:      n.conf.Severity,
+				CustomDetails: details,
+			}
+			if n.conf.Component == "" {
```
Comment (josedonizetti) [inline notify/impl.go]: You setting it on the payload if empty?

File: notify/impl.go [hunk 5]
```diff
@@ -474,17 +488,41 @@ func (n *PagerDuty) Notify(ctx context.Context, as ...*types.Alert) (bool, error
...
+				msgpayload.Component = n.conf.Component
+			}
+			if n.conf.Group == "" {
```
Comment (josedonizetti) [inline notify/impl.go]: You setting it on the payload if empty?

File: notify/impl.go [hunk 6]
```diff
@@ -474,17 +488,41 @@ func (n *PagerDuty) Notify(ctx context.Context, as ...*types.Alert) (bool, error
...
-	msg := &pagerDutyMessage{
-		ServiceKey:  tmpl(string(n.conf.ServiceKey)),
-		EventType:   eventType,
-		IncidentKey: hashKey(key),
-		Description: tmpl(n.conf.Description),
-		Details:     details,
+	msg := &pagerDutyMessage{}
+
+	if n.conf.ServiceKey != "" {
+		msg.ServiceKey = tmpl(string(n.conf.ServiceKey))
```
Comment (josedonizetti) [inline notify/impl.go]: Add some logging here saying that the v1 support is deprecated. Maybe add the user to the pagerduty faq where they day the date that they still stop supporting it.

File: config/config_test.go (hunk 2 above)
Comment (scarvel8) [reply config/config_test.go]: removing this option from the global config

File: notify/impl.go (hunk 6 above)
Comment (scarvel8) [reply notify/impl.go]: :+1:

File: notify/impl.go (hunk 4 above)
Comment (scarvel8) [reply notify/impl.go]: good catch, thanks

File: notify/impl.go (hunk 3 above)
Comment (scarvel8) [reply notify/impl.go]: Strangely, ServiceKey (which is the v1 equivalent) was being processed as a template as well in the past

File: notify/impl.go (hunk 1 above)
Comment (scarvel8) [reply notify/impl.go]: they are not

File: notify/impl.go [hunk 7]
```diff
@@ -474,17 +488,42 @@ func (n *PagerDuty) Notify(ctx context.Context, as ...*types.Alert) (bool, error
...
 
-	msg := &pagerDutyMessage{
-		ServiceKey:  tmpl(string(n.conf.ServiceKey)),
-		EventType:   eventType,
-		IncidentKey: hashKey(key),
-		Description: tmpl(n.conf.Description),
-		Details:     details,
+	msg := &pagerDutyMessage{}
+
+	if n.conf.ServiceKey != "" {
```
Comment (stuartnelson3) [inline notify/impl.go]: Could you branch here into two separate functions, something like 
```go
if n.conf.ServiceKey != "" {
	level.Info(n.logger).Log("msg", "PagerDuty v1 API will no longer be supported: https://v2.developer.pagerduty.com/v2/docs/api-v2-frequently-asked-questions")
	return n.notifyV1(ctx, as...)
else {
        return n.notifyV2(ctx, as...)
}
```

And have the same for the retry logic?

**Comment by scarvel8** [comment]
@josedonizetti retry logic added
@stuartnelson3 added the branching logic for v1 vs v2

Squashed commits

**Comment by josedonizetti** [comment]
@scarvel8 I've mistakenly asked you to not use templates for the PD configuration keys (RoutingKey, and ServiceKey), but apparently, this is a good way to keep the keys outside of the source code repository. See https://github.com/prometheus/alertmanager/issues/619

@stuartnelson3 Does it make sense?

**Comment by brian-brazil** [comment]
> this is a good way to keep the keys outside of the source code repository. See #619 (sorry :()

This is incorrect.

**Comment by brian-brazil** [comment]
To expand, we currently template things used for routing but not things which are purely used as credentials.

**Comment by josedonizetti** [comment]
Thanks for the answer @brian-brazil

**Pull request merged**